src/storage-client/src/controller.rs

```diff
@@ -1995,14 +2004,70 @@ where
             data_shard,
         } = new_metadata;
 
-        // Update in memory collection metadata.
-        let metadata = self
-            .state
-            .collections
-            .get_mut(&id)
-            .expect("id {:?} must exist");
-        metadata.collection_metadata.data_shard = data_shard;
-        metadata.collection_metadata.remap_shard = remap_shard;
+        // Update in memory collection metadata if the collection has been
+        // registered.
+        if let Some(metadata) = self.state.collections.get_mut(&id) {
+            metadata.collection_metadata.data_shard = data_shard;
+            metadata.collection_metadata.remap_shard = remap_shard;
+
+            if replace_data_shard {
```

Comment:
Yeah, this is not totally clear out of context! We have two situations that we need to support:
- Instantiating progress subsources for new sources. In this case, we will not be doing a rewrite of any of the shards.
- Instantiating progress subsources for sources that existed in previous versions of MZ. In this case, the call to `create_collection` doesn't know that this progress subsource is going to belong to an "old source." However, when the "old source" gets re-instantiated, it will see that it has a remap shard that needs to become the data shard of its progress subsource, and we call `rewrite_collection_metadata`. In this case, we're replacing the data shard.

The tension here is that when we create the progress collection, we don't know if it will belong to a new or existing source. If it belong to an existing source, we need to change the in-memory data as well as the data stored in stash.